compiler/src/middle_end/matchcomp.re

```diff
@@ -238,26 +244,59 @@ module MatchTreeCompiler = {
 
   let fold_tree = (setup, ans) =>
     List.fold_right(
-      ((name, exp), body) => AExp.let_(Nonrecursive, [(name, exp)], body),
+      (bind, body) =>
+        switch (bind) {
+        | BLet(name, exp) => AExp.let_(Nonrecursive, [(name, exp)], body)
+        | _ =>
+          failwith("match_comp: compile_tree_help: bindings were not BLet")
+        },
       setup,
       AExp.comp(ans),
     );
 
-  let rec compile_tree_help = (tree, values) => {
+  let rec compile_tree_help = (tree, values, expr, helpI) => {
     let (cur_value, rest_values) =
       switch (values) {
       | [] => failwith("Impossible (compile_tree_help): Empty value stack")
       | [hd, ...tl] => (hd, tl)
       };
     switch (tree) {
     | Leaf(i) => (Comp.imm(Imm.const(Const_int(i))), [])
+    | Guard(mb, true_tree, false_tree) =>
+      let guard =
+        switch (mb.mb_guard) {
+        | Some(guard) => guard
+        | None =>
+          failwith(
+            "Impossible: (compile_tree_help): guarded match branch contained no guard",
+          )
```

Comment:
See my earlier point... though I honestly feel this could just be `assert(false)`.

Reply:
I am generally against `assert(false)`-type things. At least a failure message like "Impossible" might get someone to open an issue with "well actually!" so we can fix it.